File: grid.py
import sys
from Queue import *

class Grid:
    """Grid Class"""

    width_ = 0
    length_ = 0
    xstart_ = 0
    ystart_ = 0
    xend_ = 0
    yend_ = 0
    grid_ = []
    visiting_ = Queue(1000)

    def __init__(self, width = 8, length = 8):
        self.width_ = width
        self.length_ = length

        self.grid_ = [[-1 for x in xrange(width)] for y in xrange(length)] 
        
    def setStart(self, x, y):
        if (x < 0 or x > self.width_):
            raise ValueError("x value of setBegin out of bounds.")
        elif (y < 0 or y > self.length_):
            raise ValueError("y value of setBegin out of bounds.")
        else:
            self.xstart_ = x
            self.ystart_ = y
            self.grid_[y][x] = 0

    def setEnd(self, x, y):
        if (x < 0 or x > self.width_):
            raise ValueError("x value of setEnd out of bounds.")
        elif (y < 0 or y > self.length_):
            raise ValueError("y value of setEnd out of bounds.")
        else:
            self.xend_ = x
            self.yend_ = y
            self.grid_[y][x] = -2

    def simulate(self):
        print("Simulating...")
        timeout = 80

        self.visiting_.put({ 'x': self.xstart_, 'y': self.ystart_ })

        searching = True
        
        # Wait for timeout, in case of bugs
        while(not self.visiting_.empty() and searching and timeout > 0):
            searching = self.simulateStep()
            timeout -= 1
        
        self.printGrid()

        if (self.visiting_.empty()):
            print("Could not find path to end!")
        elif (timeout == 0):
            print("Timeout! Failed to find end in time.")
        else:
            print("Found End! Calculating Path...")
            self.calculatePath()

    def simulateStep(self):
        id = self.visiting_.get()

        if (id['x'] == self.xend_ and id['y'] == self.yend_):
            return False
        else:
            v = self.grid_[id['y']][id['x']] + 1

            if (id['x'] > 0):
                if (self.checkValue(self.grid_[id['y']][id['x'] - 1])):
                    self.visiting_.put({'x': id['x'] - 1, 'y': id['y']})
                    self.grid_[id['y']][id['x'] - 1] = v
                elif (self.grid_[id['y']][id['x'] - 1] == -2):
                    return False

            if (id['x'] < self.width_ - 1):
                if (self.checkValue(self.grid_[id['y']][id['x'] + 1])):
                    self.visiting_.put({'x': id['x'] + 1, 'y': id['y']})
                    self.grid_[id['y']][id['x'] + 1] = v
                elif (self.grid_[id['y']][id['x'] + 1] == -2):
                    return False

            if (id['y'] > 0):
                if (self.checkValue(self.grid_[id['y'] - 1][id['x']])):
                    self.visiting_.put({'x': id['x'], 'y': id['y'] - 1})
                    self.grid_[id['y'] - 1][id['x']] = v
                elif (self.grid_[id['y'] - 1][id['x']] == -2):
                    return False

            if (id['y'] < self.length_ - 1):
                if (self.checkValue(self.grid_[id['y'] + 1][id['x']])):
                    self.visiting_.put({'x': id['x'], 'y': id['y'] + 1})
                    self.grid_[id['y'] + 1][id['x']] = v
                elif (self.grid_[id['y'] + 1][id['x']] == -2):
                    return False
            
        return True

    def printGrid(self):
        print("Printing Grid")

        for i in range(0, self.length_):
            line = ""

            for j in range(0, self.width_):
                v = self.printSymbol(self.grid_[i][j]) 
                line += v + " "
            
            print(line)

        print("-------------")

    def printSymbol(self, val):
        switch = {
            -1: "__",
            0:  "ss",
            -2: "ee",
            -3: "##"
        }
        return switch.get(val, "{0:0=2d}".format(val))

    def checkCell(self, x, y, check):
        if (check == 0 and y < self.length_ - 1):
            return self.checkValue(self.grid_[y + 1][x])
        if (check == 1 and y > 0):
            return self.checkValue(self.grid_[y - 1][x])
        if (check == 2 and x < self.width_ - 1):
            return self.checkValue(self.grid_[y][x + 1])
        if (check == 3 and x > 0):
            return self.checkValue(self.grid_[y][x - 1])

        return False

    def checkValue(self, val):
        return (val == -1)

    def setBlock(self, x, y):
        if (x < 0 or x > self.width_):
            raise ValueError("x value of setBlock out of bounds.")
        elif (y < 0 or y > self.length_):
            raise ValueError("y value of setBlock out of bounds.")
        else:
            self.grid_[x][y] = -3

    def calculatePath(self):
        print("Calculating Path.")
        path = []

        x = self.xend_
        y = self.yend_
        min = sys.maxint

        processing = True
        while (processing):
            xmin = x
            ymin = y

            if (x > 0 and self.grid_[y][x - 1] >= 0 and self.grid_[y][x - 1] < min):
                path.append({'x': x - 1, 'y': y})
                min = self.grid_[y][x - 1]
                xmin = x - 1
                ymin = y

            if (x < self.width_ - 1 and self.grid_[y][x + 1] >= 0 and self.grid_[y][x + 1] < min):
                min = self.grid_[y][x + 1]
                xmin = x + 1
                ymin = y

            if (y > 0 and self.grid_[y - 1][x] >= 0 and self.grid_[y - 1][x] < min):
                min = self.grid_[y - 1][x]
                xmin = x
                ymin = y - 1

            if (y < self.length_ - 1 and self.grid_[y + 1][x] >= 0 and self.grid_[y + 1][x] < min):
                min = self.grid_[y + 1][x]
                xmin = x
                ymin = y + 1
            
            if (min == 0):
                processing = False
            else:
                path.append({'x': xmin, 'y': ymin})

                x = xmin
                y = ymin

        for i in range(0, self.length_):
            line = ""
            for j in range(0, self.width_):
                found = False
                for k in range(0, len(path)):
                    id = path[k]

                    if (id['x'] == j and id['y'] == i):
                        found = True
                        break

                if (found):
                    line += "x "
                elif (self.grid_[i][j] == 0):
                    line += "s "
                elif (self.grid_[i][j] == -2):
                    line += "e "
                elif (self.grid_[i][j] == -3):
                    line += "# "
                else:
                    line += "_ "

            print(line)

File: Grid.cpp
#include "Grid.h"
#include <stdexcept>
#include <iostream>
#include <string>
#include <iomanip>

const bool DEBUG_PATH = false;

CellID::CellID(int x, int y, int dist) {
    this->x = x;
    this->y = y;
    this->dist = dist;
}

Grid::Grid(int w, int l) : width_(w), length_(l) {
    if (w < 0 || l < 0) {
        throw std::runtime_error("Invalid Size for Grid.");
    }

    grid_ = new Cell *[l];

    for (int i = 0; i < l; ++i) {
        grid_[i] = new Cell[w];

        for (int j = 0; j < w; ++j) {
            grid_[i][j] = -1;
        }
    }
}

void Grid::setStart(int x, int y) {
    xstart_ = x;
    ystart_ = y;
    grid_[y][x] = 0;
}

void Grid::setEnd(int x, int y) {
    xend_ = x;
    yend_ = y;
    grid_[y][x] = -2;
}

void Grid::setBlock(int x, int y) {
    grid_[y][x] = -3;
}

bool Grid::checkValue(Cell val) {
    return (val == -1);
}

bool Grid::checkCell(int x, int y, Check check) {
    switch (check) {
        case CHECK_UP:
            if (y < length_ - 1)
                return checkValue(grid_[y + 1][x]);
        case CHECK_DOWN:
            if (y > 0)
                return checkValue(grid_[y - 1][x]);
        case CHECK_RIGHT:
            if (x < width_ - 1)
                return checkValue(grid_[y][x + 1]);
        case CHECK_LEFT:
            if (x > 0)
                return checkValue(grid_[y][x - 1]);
    }

    return false;
}

bool Grid::simulateStep() {
    CellID id;
    id = visiting_.front();
    visiting_.pop();

    if (id.x == xend_ && id.y == yend_) {
        return false;
    }
    else {
        int v = grid_[id.y][id.x] + 1;

        if (id.x > 0) {
            if (checkValue(grid_[id.y][id.x - 1])) {
                visiting_.emplace(id.x - 1, id.y, v);
                grid_[id.y][id.x - 1] = v;
            }
            else if (grid_[id.y][id.x - 1] == -2) {
                return false;
            }
        }

        if (id.x < width_ - 1) {
            if (checkValue(grid_[id.y][id.x + 1])) {
                visiting_.emplace(id.x + 1, id.y, v);
                grid_[id.y][id.x + 1] = v;
            }
            else if (grid_[id.y][id.x + 1] == -2) {
                return false;
            }
        }

        if (id.y > 0) {
            if (checkValue(grid_[id.y - 1][id.x])) {
                visiting_.emplace(id.x, id.y - 1, v);
                grid_[id.y - 1][id.x] = v;
            }
            else if (grid_[id.y - 1][id.x] == -2) {
                return false;
            }
        }

        if (id.y < length_ - 1) {
            if (checkValue(grid_[id.y + 1][id.x])) {
                visiting_.emplace(id.x, id.y + 1, v);
                grid_[id.y + 1][id.x] = v;
            }
            else if (grid_[id.y + 1][id.x] == -2) {
                return false;
            }
        }
    }

    if (DEBUG_PATH) {
        printGrid();
        std::cout << "_______________\n";
    }

    return true;
}

void Grid::calculatePath() {
    visiting_.empty();

    std::vector<CellID> path;

    int x = xend_;
    int y = yend_;
    int min = INT32_MAX;
    while (true) {
        int xmin = x;
        int ymin = y;

        if (x > 0 && grid_[y][x - 1] >= 0 && grid_[y][x - 1] < min) {
            path.emplace_back(x - 1, y, 0);
            min = grid_[y][x - 1];
            xmin = x - 1;
            ymin = y;
        }

        if (x < width_ - 1 && grid_[y][x + 1] >= 0 && grid_[y][x + 1] < min) {
            min = grid_[y][x + 1];
            xmin = x + 1;
            ymin = y;
        }

        if (y > 0 && grid_[y - 1][x] >= 0 && grid_[y - 1][x] < min) {
            min = grid_[y - 1][x];
            xmin = x;
            ymin = y - 1;
        }

        if (y < length_ - 1 && grid_[y + 1][x] >= 0 && grid_[y + 1][x] < min) {
            min = grid_[y + 1][x];
            xmin = x;
            ymin = y + 1;
        }
        
        if (min == 0) {
            break;
        }

        path.emplace_back(xmin, ymin, 0);

        x = xmin;
        y = ymin;
    }

    for (int i = 0; i < length_; ++i) {
        for (int j = 0; j < width_; ++j) {
            bool found = false;
            for (int k = 0; k < path.size(); ++k) {
                CellID id = path[k];

                if (id.x == j && id.y == i) {
                    found = true;
                    break;
                }
            }

            if (found) {
                std::cout << "x ";
            }
            else if (grid_[i][j] == 0) {
                std::cout << "s ";
            }
            else if (grid_[i][j] == -2) {
                std::cout << "e ";
            }
            else if (grid_[i][j] == -3) {
                std::cout << "# ";
            }
            else {
                std::cout << "_ ";
            }
        }

        std::cout << "\n";
    }
}

void Grid::simulate() {
    int timeout = 80;

    visiting_.emplace(xstart_, ystart_, 0);

    // Wait for timeout, in case of bugs
    while(visiting_.size() > 0 && simulateStep() && --timeout > 0);

    if (!DEBUG_PATH) {
        printGrid();
    }

    if (visiting_.size() == 0) {
        std::cout << "Could not find path to end!\n";
    }
    else if (timeout == 0) {
        std::cout << "Timeout! Failed to find end in time.\n";
    }
    else {
        std::cout << "Found End! Tacking...\n";

        calculatePath();
    }
}

std::string Grid::printSymbol(Cell val) {
    switch(val) {
    default:
        return std::to_string(val);
    case -1:
        return "__";
    case 0:
        return "ss";
    case -2:
        return "ee";
    case -3:
        return "##";
    }
}

void Grid::printGrid() {
    for (int i = 0; i < length_; ++i) {
        for (int j = 0; j < width_; ++j) {
            std::cout << std::setw(2) << printSymbol(grid_[i][j]) << " ";
        }
        std::cout << "\n";
    }
}

Grid::~Grid() {
    if (grid_ != nullptr) {
        /*for (int i = 0; i < length_; ++i) {
            delete[] grid_[i];
        }*/

        delete[] grid_;
    }
}